genericjpa/db/src/main/java/org/hibernate/search/genericjpa/db/events/index/impl/IndexUpdater.java

```diff
@@ -0,0 +1,320 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.genericjpa.db.events.index.impl;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.hibernate.search.backend.spi.SingularTermDeletionQuery;
+import org.hibernate.search.backend.spi.Work;
+import org.hibernate.search.backend.spi.WorkType;
+import org.hibernate.search.bridge.FieldBridge;
+import org.hibernate.search.bridge.StringBridge;
+import org.hibernate.search.engine.ProjectionConstants;
+import org.hibernate.search.engine.integration.impl.ExtendedSearchIntegrator;
+import org.hibernate.search.engine.metadata.impl.DocumentFieldMetadata;
+import org.hibernate.search.genericjpa.db.EventType;
+import org.hibernate.search.genericjpa.db.events.UpdateConsumer.UpdateEventInfo;
+import org.hibernate.search.genericjpa.entity.EntityProvider;
+import org.hibernate.search.genericjpa.entity.ReusableEntityProvider;
+import org.hibernate.search.genericjpa.exception.SearchException;
+import org.hibernate.search.genericjpa.factory.Transaction;
+import org.hibernate.search.genericjpa.metadata.impl.RehashedTypeMetadata;
+import org.hibernate.search.genericjpa.util.NamingThreadFactory;
+import org.hibernate.search.query.engine.spi.EntityInfo;
+import org.hibernate.search.query.engine.spi.HSQuery;
+
+/**
+ * This class is the "glue" between the updating mechanism and the actual
+ * Hibernate-Search index. It consumes Events coming from the AsyncUpdateSource and updates the Hibernate-Search index
+ * accordingly
+ *
+ * @author Martin Braun
+ */
+public final class IndexUpdater {
+
+	// TODO: think of a clever way of doing batching here
+	// or maybe leave it as it is
+
+	// TODO: unit test this with several batches
+
+	private static final Logger LOGGER = Logger.getLogger( IndexUpdater.class.getName() );
```

Comment:
Also, I should probably replace all of these with the default way of logging in Hibernate Search.